### ISSUE TITLE
Prometheus configure with EFS on AWS

### ISSUE THREAD

**Issue by Denusdv** [prometheus-operator/prometheus-operator#3150]
**Try to config Prometheus to with against EFS on AWS**

My goal is to use PersistentVolume of nfs type
I create EFS on AWS and  want to mount it to Prometheus work with it.
The problem I face with is that Prometheus keeps use previously created ebs and kind of ignore the new configuration, the yaml files are attached.

* Prometheus Operator version: v0.34.0

    `quay.io/coreos/prometheus-operator:v0.34.0`
    

* Kubernetes version information:

   Deploy the Prometheus on EKS of AWS

* Manifests:

PersistentVolume
```
apiVersion: v1
kind: PersistentVolume
metadata:
  name: prometheus-pv-name
  labels:
    app: prometheus-pv-name
spec:
  capacity:
    storage: 50Gi
  volumeMode: Filesystem
  accessModes:
    - ReadWriteMany # required
  storageClassName: aws-efs
  nfs:
    server: fs-315b55fa.efs.eu-west-1.amazonaws.com
    path: /
```
Prometheus
```
apiVersion: monitoring.coreos.com/v1
kind: Prometheus
metadata:
  labels:
    app: prometheus
    prometheus: bbas-prometheus
  name: bbas-prometheus
  namespace: monitoring
spec:
  securityContext:
    runAsUser: 1000
    fsGroup: 2000
    runAsNonRoot: true
  alerting:
    alertmanagers:
    - name: alertmanager-main
      namespace: monitoring
      port: web
  baseImage: quay.io/prometheus/prometheus
  nodeSelector:
    pod-type: general
  logLevel: info
  paused: false
  replicas: 1
  retention: 90d
  routePrefix: /
  ruleSelector:
    matchLabels:
      prometheus: bbas-prometheus
      role: alert-rules
  serviceAccountName: prometheus-k8s
  serviceMonitorNamespaceSelector: {}
  serviceMonitorSelector:
    matchExpressions:
    - key: serviceapp
      operator: Exists
  storage:
    volumeClaimTemplate:
      spec:
        resources:
          requests:
            storage: 50Gi
      selector:
        matchLabels:
          app: prometheus-pv-name

```

* Prometheus Operator Logs:

**kubectl describe pod prometheus-bbas-prometheus-0 -n monitoring**

```
Name:           prometheus-bbas-prometheus-0
Namespace:      monitoring
Priority:       0
Node:           ip-172-31-5-91.eu-west-1.compute.internal/172.31.5.91
Start Time:     Thu, 16 Apr 2020 21:53:12 +0300
Labels:         app=prometheus
                controller-revision-hash=prometheus-bbas-prometheus-6dc4d856c8
                prometheus=bbas-prometheus
                statefulset.kubernetes.io/pod-name=prometheus-bbas-prometheus-0
Annotations:    kubernetes.io/psp: eks.privileged
Status:         Running
IP:             172.31.7.126
IPs:            <none>
Controlled By:  StatefulSet/prometheus-bbas-prometheus
Containers:
  prometheus:
    Container ID:  docker://9e76741148ed3728883ce60d368217cbda81a1ade8339c518f369b5e39d7c76b
    Image:         quay.io/prometheus/prometheus:v2.7.1
    Image ID:      docker-pullable://quay.io/prometheus/prometheus@sha256:f21aa9225a1ba630ea1671c4c7d8595aa87a6e89bbb6525c56703ed317ae9226
    Port:          9090/TCP
    Host Port:     0/TCP
    Args:
      --web.console.templates=/etc/prometheus/consoles
      --web.console.libraries=/etc/prometheus/console_libraries
      --config.file=/etc/prometheus/config_out/prometheus.env.yaml
      --storage.tsdb.path=/prometheus
      --storage.tsdb.retention.time=90d
      --web.enable-lifecycle
      --storage.tsdb.no-lockfile
      --web.route-prefix=/
    State:       Running
      Started:   Thu, 16 Apr 2020 21:53:32 +0300
    Last State:  Terminated
      Reason:    Error
      Message:   2020-04-16T18:53:31.258927093Z caller=main.go:305 fd_limits="(soft=65536, hard=65536)"
level=info ts=2020-04-16T18:53:31.258950973Z caller=main.go:306 vm_limits="(soft=unlimited, hard=unlimited)"
level=info ts=2020-04-16T18:53:31.259617326Z caller=main.go:620 msg="Starting TSDB ..."
level=info ts=2020-04-16T18:53:31.259768615Z caller=web.go:416 component=web msg="Start listening for connections" address=0.0.0.0:9090
level=info ts=2020-04-16T18:53:31.265952672Z caller=main.go:635 msg="TSDB started"
level=info ts=2020-04-16T18:53:31.265992507Z caller=main.go:695 msg="Loading configuration file" filename=/etc/prometheus/config_out/prometheus.env.yaml
level=info ts=2020-04-16T18:53:31.26605034Z caller=main.go:489 msg="Stopping scrape discovery manager..."
level=info ts=2020-04-16T18:53:31.2660623Z caller=main.go:503 msg="Stopping notify discovery manager..."
level=info ts=2020-04-16T18:53:31.26607055Z caller=main.go:525 msg="Stopping scrape manager..."
level=info ts=2020-04-16T18:53:31.266079401Z caller=main.go:499 msg="Notify discovery manager stopped"
level=info ts=2020-04-16T18:53:31.266102509Z caller=main.go:485 msg="Scrape discovery manager stopped"
level=info ts=2020-04-16T18:53:31.266121533Z caller=main.go:519 msg="Scrape manager stopped"
level=info ts=2020-04-16T18:53:31.26617673Z caller=manager.go:736 component="rule manager" msg="Stopping rule manager..."
level=info ts=2020-04-16T18:53:31.266210058Z caller=manager.go:742 component="rule manager" msg="Rule manager stopped"
level=info ts=2020-04-16T18:53:31.268642424Z caller=notifier.go:521 component=notifier msg="Stopping notification manager..."
level=info ts=2020-04-16T18:53:31.268665016Z caller=main.go:679 msg="Notifier manager stopped"
level=error ts=2020-04-16T18:53:31.268787519Z caller=main.go:688 err="error loading config from \"/etc/prometheus/config_out/prometheus.env.yaml\": couldn't load configuration (--config.file=\"/etc/prometheus/config_out/prometheus.env.yaml\"): open /etc/prometheus/config_out/prometheus.env.yaml: no such file or directory"

      Exit Code:    1
      Started:      Thu, 16 Apr 2020 21:53:31 +0300
      Finished:     Thu, 16 Apr 2020 21:53:31 +0300
    Ready:          True
    Restart Count:  1
    Liveness:       http-get http://:web/-/healthy delay=0s timeout=3s period=5s #success=1 #failure=6
    Readiness:      http-get http://:web/-/ready delay=0s timeout=3s period=5s #success=1 #failure=120
    Environment:    <none>
    Mounts:
      /etc/prometheus/certs from tls-assets (ro)
      /etc/prometheus/config_out from config-out (ro)
      /etc/prometheus/rules/prometheus-bbas-prometheus-rulefiles-0 from prometheus-bbas-prometheus-rulefiles-0 (rw)
      /prometheus from prometheus-bbas-prometheus-db (rw,path="prometheus-db")
      /var/run/secrets/kubernetes.io/serviceaccount from prometheus-k8s-token-nkq9r (ro)
  prometheus-config-reloader:
    Container ID:  docker://33d1dfa9eeeaaa81a616254a8c1dd5b89bc4a1a417224329fd857351d0e7fa09
    Image:         quay.io/coreos/prometheus-config-reloader:v0.34.0
    Image ID:      docker-pullable://quay.io/coreos/prometheus-config-reloader@sha256:459bc1b1fc51420cd244b082fe14444f6d8596997c2e40f52c08d1f5a09ab860
    Port:          <none>
    Host Port:     <none>
    Command:
      /bin/prometheus-config-reloader
    Args:
      --log-format=logfmt
      --reload-url=http://localhost:9090/-/reload
      --config-file=/etc/prometheus/config/prometheus.yaml.gz
      --config-envsubst-file=/etc/prometheus/config_out/prometheus.env.yaml
    State:          Running
      Started:      Thu, 16 Apr 2020 21:53:31 +0300
    Ready:          True
    Restart Count:  0
    Limits:
      cpu:     100m
      memory:  25Mi
    Requests:
      cpu:     100m
      memory:  25Mi
    Environment:
      POD_NAME:  prometheus-bbas-prometheus-0 (v1:metadata.name)
    Mounts:
      /etc/prometheus/config from config (rw)
      /etc/prometheus/config_out from config-out (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from prometheus-k8s-token-nkq9r (ro)
  rules-configmap-reloader:
    Container ID:  docker://4e026b80a02f5fae2f80114905672032c6ac0a490f55223e7b4bdd4c17c23ef2
    Image:         quay.io/coreos/configmap-reload:v0.0.1
    Image ID:      docker-pullable://quay.io/coreos/configmap-reload@sha256:e2fd60ff0ae4500a75b80ebaa30e0e7deba9ad107833e8ca53f0047c42c5a057
    Port:          <none>
    Host Port:     <none>
    Args:
      --webhook-url=http://localhost:9090/-/reload
      --volume-dir=/etc/prometheus/rules/prometheus-bbas-prometheus-rulefiles-0
    State:          Running
      Started:      Thu, 16 Apr 2020 21:53:31 +0300
    Ready:          True
    Restart Count:  0
    Limits:
      cpu:     100m
      memory:  25Mi
    Requests:
      cpu:        100m
      memory:     25Mi
    Environment:  <none>
    Mounts:
      /etc/prometheus/rules/prometheus-bbas-prometheus-rulefiles-0 from prometheus-bbas-prometheus-rulefiles-0 (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from prometheus-k8s-token-nkq9r (ro)
Conditions:
  Type              Status
  Initialized       True
  Ready             True
  ContainersReady   True
  PodScheduled      True
Volumes:
  prometheus-bbas-prometheus-db:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  prometheus-bbas-prometheus-db-prometheus-bbas-prometheus-0
    ReadOnly:   false
  config:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  prometheus-bbas-prometheus
    Optional:    false
  tls-assets:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  prometheus-bbas-prometheus-tls-assets
    Optional:    false
  config-out:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:
    SizeLimit:  <unset>
  prometheus-bbas-prometheus-rulefiles-0:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      prometheus-bbas-prometheus-rulefiles-0
    Optional:  false
  prometheus-k8s-token-nkq9r:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  prometheus-k8s-token-nkq9r
    Optional:    false
QoS Class:       Burstable
Node-Selectors:  pod-type=general
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type    Reason                  Age                From                                                Message
  ----    ------                  ----               ----                                                -------
  Normal  Scheduled               17m                default-scheduler                                   Successfully assigned monitoring/prometheus-bbas-prometheus-0 to ip-172-31-5-91.eu-west-1.compute.internal
  Normal  SuccessfulAttachVolume  17m                attachdetach-controller                             AttachVolume.Attach succeeded for volume "pvc-75e0b418-8013-11ea-b9b5-024ebb604748"
  Normal  Pulled                  16m                kubelet, ip-172-31-5-91.eu-west-1.compute.internal  Container image "quay.io/coreos/configmap-reload:v0.0.1" already present on machine
  Normal  Created                 16m                kubelet, ip-172-31-5-91.eu-west-1.compute.internal  Created container rules-configmap-reloader
  Normal  Started                 16m                kubelet, ip-172-31-5-91.eu-west-1.compute.internal  Started container rules-configmap-reloader
  Normal  Pulled                  16m                kubelet, ip-172-31-5-91.eu-west-1.compute.internal  Container image "quay.io/coreos/prometheus-config-reloader:v0.34.0" already present on machine
  Normal  Created                 16m                kubelet, ip-172-31-5-91.eu-west-1.compute.internal  Created container prometheus-config-reloader
  Normal  Started                 16m                kubelet, ip-172-31-5-91.eu-west-1.compute.internal  Started container prometheus-config-reloader
  Normal  Pulled                  16m (x2 over 16m)  kubelet, ip-172-31-5-91.eu-west-1.compute.internal  Container image "quay.io/prometheus/prometheus:v2.7.1" already present on machine
  Normal  Created                 16m (x2 over 16m)  kubelet, ip-172-31-5-91.eu-west-1.compute.internal  Created container prometheus
  Normal  Started                 16m (x2 over 16m)  kubelet, ip-172-31-5-91.eu-west-1.compute.internal  Started container prometheus
```

please help me to figure out what I'm missing.

**Comment by brancz** [comment]
You need to delete the whole setup including the PVs and PVCs, and recreate.

**Comment by Denusdv** [comment]
I did it a many times. it didn't help.

**Comment by brancz** [comment]
Prometheus' volume must not be read write many

```
    - ReadWriteMany # required
```

The storage must not be shared. The Prometheus Operator will force the requested volume to be ReadWriteOnce.

**Comment by Denusdv** [comment]
Hi Frederic, thanks for looking and trying to help me. I still face the same result even after the 
**accessModes** is set to - ReadWriteOnce.

My current PV is 

```
apiVersion: v1
kind: PersistentVolume
metadata:
  name: prometheus-pv-name
  labels:
    app: prometheus-pv-name
spec:
  capacity:
    storage: 50Gi
  volumeMode: Filesystem
  accessModes:
    - ReadWriteOnce
  nfs:
    server: fs-315b55fa.efs.eu-west-1.amazonaws.com
    path: /
```

The storage section in the Prometheus is

```
storage:
    volumeClaimTemplate:
      spec:
        resources:
          requests:
            storage: 50Gi
      selector:
        matchLabels:
          app: prometheus-pv-name
```

the storage points to the PV rather than PVC, although I following the instructions where I should specify PV.

I created the Prometheus completely including PV. 
The result I got, the Prometheus deployment created 

```
kubernetes-dynamic-pvc-8c567306-80ce-11ea-b9b5-024ebb604748 | vol-005068a5973786dc6 | 50 GiB | gp2 | 150 |   | April 17, 2020 at 8:12:15 PM UTC+3 | eu-west-1b

```

meaning, we got gp2 storage of 50GiB, although the PV points to the EFS on fs-315b55fa.efs.eu-west-1.amazonaws.com.

Seems like the deployment just ignores the PV and using some default configuration to create 50Gi on gp2.

What I found is that we have the default **StorageClass**
```
kubectl get StorageClass -n monitoring
NAME            PROVISIONER             AGE
gp2 (default)   kubernetes.io/aws-ebs   5d21h 
```

Could it be related to the problem I'm struggling.
Thanks

**Comment by brancz** [comment]
I think this has to do with the way storage classes work, as you don't specify one in your volumeclaimtemplate, it uses the default one. I'd try asking the workload API folks on Kubernetes as this is really about the statefulset.

On another note, [NFS volumes are unsupported by Prometheus](https://prometheus.io/docs/prometheus/latest/storage/#operational-aspects), you should not use them.

**Comment by kiuka** [comment]
@brancz why would it be unsupported? 

"NFS is only potentially POSIX, most implementations are not."

**Comment by brancz** [comment]
Prometheus' storage relies on certain semantics, which are not POSIX compliant in many NFS implementations. With this significant unrecoverable storage corruptions can happen (and have happened).

**Comment by cah-andrewfitzgerald** [comment]
For what it's worth, the [Amazon EFS docs](https://docs.aws.amazon.com/efs/latest/ug/creating-using.html) state that it is POSIX compliant.
<img width="973" alt="Screen Shot 2020-05-12 at 5 07 22 PM" src="https://user-images.githubusercontent.com/13521755/81745567-130d2580-9473-11ea-993d-5945632ee2b4.png">

**Comment by brancz** [comment]
That does increase confidence. I would still be cautious, but just because I've been burned :)

**Comment by pinguo-yaojun** [comment]
> That does increase confidence. I would still be cautious, but just because I've been burned :)

Did you find the way to use EFS in prometheus?

**Comment by wmcnamee-tunein** [comment]
Also interested if anyone has successfully used EFS with Prometheus.

**Comment by ghostsquad** [comment]
also curious if anyone has gotten this to work and/or had any problems with EFS

**Comment by chris-isaak** [comment]
I have a very basic Prometheus, configured to pull from cAdvisor in a POC using Fargate and EFS.  I had to create the prometheus.yml file before launch for it to work.  I do wish I knew why the docker image is not writing to the EFS itself.  Again my setup is very basic. Using POSIX permission 755 for the share.

**Comment by roidelapluie** [comment]
Hello,

For anyone stumbling into this issue: we have lots of users reporting data corruption with EFS. It should be avoided.

**Comment by erickpeirson** [comment]
> For anyone stumbling into this issue: we have lots of users reporting data corruption with EFS. It should be avoided.

@roidelapluie Are you able to share any more info about these reports? I note the discussion of non-POSIX filesystems in the docs, and Google turns up anecdotes, but would love to read more about the specific cases you're seeing (if possible).

**Comment by roidelapluie** [comment]
> > For anyone stumbling into this issue: we have lots of users reporting data corruption with EFS. It should be avoided.
> 
> @roidelapluie Are you able to share any more info about these reports? I note the discussion of non-POSIX filesystems in the docs, and Google turns up anecdotes, but would love to read more about the specific cases you're seeing (if possible).

https://github.com/prometheus/prometheus/issues/5617

**Comment by coderanger** [comment]
I think we can probably close this out, I don't think the operator has any issues here, it uses whatever volume claim template is specified (other than requiring it be RWO). The goodness or badness of using NFS with Prometheus is an upstream concern.

**Comment by daniel-fulgido** [comment]
Unfortunately Prometheus does not work with NFS, especially with AWS EFS :(
https://prometheus.io/docs/prometheus/latest/storage/

**Comment by joseeden** [comment]
Hey, has anyone able to make EFS work as Prometheus storage?
I'm specifically trying to deploy Prometheus on an EKS cluster. I already installed the EFS driver and was trying to install Prometheus via Helm but I'm also facing the same issues that others have encountered.

I've found this link that says others have successfully implemented it although no supporting instructions were provided. 
https://www.reddit.com/r/PrometheusMonitoring/comments/gjrolv/using_efs_for_prometheusk8s_storage/

I might just switch to installing EBS driver/EBS as persistent volume if I can't find any sources on how to implement EFS as PV.
